### PR TITLE
Support message, nameFieldLabel and showsTagField for macOS dialog

### LINKED ISSUE #7783: Support setMessage for macOS dialog
NSSavePanel (and NSOpenPanel, as a subclass) supports a `message` (via `setMessage:message` in Obj-C), which allows a message to be displayed even when the panel is attached to a parent. Right now, `title` is useless for attached dialogs.

### PR DESCRIPTION
**Pull request by yamgent**
Fixes #7783.

showSaveDialog() now supports additional attributes for setting the `message`, `nameFieldLabel` and `showsTagField`.

showOpenDialog() also supports `message` (`nameFieldLabel` and `showsTagField` doesn't seem to change the dialog box though).

```javascript
  const {dialog} = require('electron');
  dialog.showSaveDialog({
    title: 'Save File',
    message: 'Select a location to save special file.',
    nameFieldLabel: 'Special Name',
    showsTagField: true
  });
```

screen shot 2017-02-02 at 12 05 05 am

### CODE REVIEW

File: lib/browser/api/dialog.js [hunk 1]
```diff
@@ -159,10 +168,27 @@ module.exports = {
...
+    }
+
+    if (nameFieldLabel == null) {
+      nameFieldLabel = ''
+    } else if (typeof nameFieldLabel !== 'string') {
+      throw new TypeError('Name field label must be a string')
+    }
+
+    if (showsTagField == null) {
+      showsTagField = false
```
Comment (kevinsawicki) [inline lib/browser/api/dialog.js]: Looks like the default OS value of this is `true` if unset so I think this should default to `true` to not break backwards-compatibility.

https://developer.apple.com/reference/appkit/nssavepanel/1525589-showstagfield?language=objc

File: atom/browser/ui/file_dialog_mac.mm [hunk 2]
```diff
@@ -144,10 +158,13 @@ void ShowOpenDialog(atom::NativeWindow* parent_window,
...
                     const Filters& filters,
                     int properties,
+                    const std::string& message,
                     const OpenDialogCallback& c) {
   NSOpenPanel* dialog = [NSOpenPanel openPanel];
 
-  SetupDialog(dialog, title, button_label, default_path, filters);
+  SetupDialog(dialog, title, button_label, default_path, filters, message,
+          // NSOpenPanel does not support name_field_label and shows_tag_field
+          "", false);
```
Comment (kevinsawicki) [inline atom/browser/ui/file_dialog_mac.mm]: The alignment here appear off, can you align this with the `dialog` argument on line 165?

File: atom/browser/ui/file_dialog_mac.mm [hunk 3]
```diff
@@ -123,11 +134,14 @@ bool ShowOpenDialog(atom::NativeWindow* parent_window,
...
                     int properties,
+                    const std::string& message,
                     std::vector<base::FilePath>* paths) {
   DCHECK(paths);
   NSOpenPanel* dialog = [NSOpenPanel openPanel];
 
-  SetupDialog(dialog, title, button_label, default_path, filters);
+  SetupDialog(dialog, title, button_label, default_path, filters, message,
+          // NSOpenPanel does not support name_field_label and shows_tag_field
+          "", false);
```
Comment (kevinsawicki) [inline atom/browser/ui/file_dialog_mac.mm]: Looks like the alignment here is off, can you align this with the `dialog` argument on line 142?

File: atom/browser/api/atom_api_dialog.cc [hunk 4]
```diff
@@ -92,6 +93,9 @@ void ShowSaveDialog(const std::string& title,
                     const std::string& button_label,
                     const base::FilePath& default_path,
                     const file_dialog::Filters& filters,
+                    const std::string& message,
+                    const std::string& name_field_label,
+                    const bool& shows_tag_field,
```
Comment (kevinsawicki) [inline atom/browser/api/atom_api_dialog.cc]: For primitives like `bool` and `int`, we don't use references or `const`, so this should just be `bool show_tags_field`.

File: lib/browser/api/dialog.js [hunk 5]
```diff
@@ -80,7 +80,8 @@ module.exports = {
       }
     }
 
-    let {buttonLabel, defaultPath, filters, properties, title} = options
+    let {buttonLabel, defaultPath, filters, properties, title,
+      message} = options
```
Comment (kevinsawicki) [inline lib/browser/api/dialog.js]: This is fine to keep on a single line, we don't enforce hard line length limits in JS.

File: lib/browser/api/dialog.js [hunk 6]
```diff
@@ -159,10 +168,27 @@ module.exports = {
       filters = []
     }
 
+    if (message == null) {
+      message = ''
+    } else if (typeof message !== 'string') {
+      throw new TypeError('Message must be a string')
```
Comment (kevinsawicki) [inline lib/browser/api/dialog.js]: Can you add an assertion for this error in `spec/api-dialog-spec.js`?

File: lib/browser/api/dialog.js [hunk 7]
```diff
@@ -159,10 +168,27 @@ module.exports = {
...
+    }
+
+    if (nameFieldLabel == null) {
+      nameFieldLabel = ''
+    } else if (typeof nameFieldLabel !== 'string') {
+      throw new TypeError('Name field label must be a string')
```
Comment (kevinsawicki) [inline lib/browser/api/dialog.js]: Can you add an assertion for this error in `spec/api-dialog-spec.js`?

**Comment by yamgent** [comment]
Thanks for your review! I will make the changes asap.

**Comment by yamgent** [comment]
@kevinsawicki I rebased on top of the latest master and incorporated all of your suggested changes.

**Comment by kevinsawicki** [comment]
Thanks for this @yamgent, I merged #8623 into this branch which added a `DialogSettings` helper object so the method signatures in `file_dialog.h` no longer needed to change when we add new options and options you added could just be added to that struct and then used directly in `file_dialog.mm`.

Great job on this 👍 🚀

**Pull request merged**